lib/domain.js

```diff
@@ -12,26 +9,43 @@ var getRulesForTld = require('./tld-rules.js');
  * @param {string} host
  * @return {String}
  */
-module.exports = function getDomain (allRules, validHosts, host) {
-  var domain = null, hostTld, rules, rule;
+module.exports = function getDomain(allRules, validHosts, host, isHostClean) {
   var _validHosts = validHosts || [];
+  var cleanHost = cleanHostValue(host, isHostClean);
 
-  if (isValid(_validHosts, host) === false) {
+  if (isValid(_validHosts, cleanHost) === false) {
     return null;
   }
 
-  host = cleanHostValue(host);
-  hostTld = extractTldFromHost(host);
-  rules = getRulesForTld(allRules, hostTld, new Rule({"firstLevel": hostTld, "isHost": _validHosts.indexOf(hostTld) !== -1}));
-  rule = getCandidateRule(host, rules);
+  // Check if `host` ends with '.' followed by one host specified in validHosts.
+  for (var i = 0; i < validHosts.length; i++) {
+    var vhost = validHosts[i];
+    if (cleanHost.indexOf(vhost) === (cleanHost.length - vhost.length) && (
+          cleanHost.length === vhost.length ||
+          cleanHost[cleanHost.length - vhost.length - 1] === '.')) {
+      return vhost;
+    }
+  }
 
-  if (rule === null) {
+  var suffix = getPublicSuffix(allRules, cleanHost, true);
+  if (suffix === null) {
+    // TODO - shouldn't it be null?
+    // Otherwise 'should return the known valid host' fails
+    // return cleanHost;
     return null;
   }
 
-  host.replace(new RegExp(rule.getPattern()), function (m, d) {
-    domain = d;
-  });
+  if (suffix.length === cleanHost.length) {
```

Comment:
What does this condition refers to? If the suffix is equal to the cleanhost, then there is no domain to return? (I wonder why the check is made upon the `length` rather than the value of the variables).

Reply:
Yes exactly, if the suffix has the same length of the host itself then they are equal (and we save a string comparison), so there is no domain to return (i.e.: `cleanHost` is just a valid public suffix).